src/main/haskell/kore/src/Kore/Step/Step.hs

```diff
@@ -65,16 +69,96 @@ import qualified Kore.Step.Strategy as Strategy
 
 {- | A strategy primitive: a rewrite rule or builtin simplification step.
  -}
-data Prim rewrite = Simplify | Rewrite !rewrite
+data Prim rewrite =
```

Comment:
When I wrote the strategy module, I imagined that we would use different "primitive rule" types for different modes of execution, and I think we could really benefit from that here. This could be much simpler if the one-path proof strategy used its own type different from the one used for concrete execution:
~~~.hs
data Prim rewrite = Simplify | Split ![rewrite]
~~~
The corresponding transition rule (`applySplit` vs `applyRewrite`) would apply each argument of `Split` to the remainders of the previous patterns. The `NonEmptyRemainderFail`, `CancelBottom` and `CancelRemainder` cases would not need to exist to preserve the concrete execution behavior because nothing about concrete execution would change. There wouldn't need to be an unusual use of `Strategy.sequence` (see my other comment). We would not have to introduce `StrategyPattern` or do extra post-processing of the execution tree.